### PR TITLE
Add more HtmlUtils APIs

### PR DESCRIPTION
**Pull request by andy-armstrong**

### CODE REVIEW

**Comment by edx-webhook** [comment]
## PR Cover Letter
- [ ] A detailed description of the changes in the body of the PR
- [ ] If available, a link to the relevant JIRA ticket in the body of the PR
  - `[PROJ-XXXX](https://openedx.atlassian.net/browse/PROJ-XXXX)`

## Sandbox
- [ ] Build a sandbox for your branch and add a link here

## Testing
- [ ] i18n
- [ ] RTL
- [ ] Unit, integration, acceptance tests as appropriate
- [ ] Analytics
- [ ] Performance
- [ ] Database migrations are backwards-compatible

## Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [ ] Code review #1
- [ ] Code review #2
- [ ] Docs review (required for UI strings/error messages)
- [ ] UX review
- [ ] Accessibility review
- [ ] Product review

## Documentation considerations
- [ ] Are there user visible changes? If so, list who is affected, and include a pointer to the relevant code.
- [ ] Should this be mentioned in the release notes? If so, make sure that you included a description.

## DevOps assistance
If it's needed, or even if you suspect it might be needed, do any/all these things as early as possible. Don't surprise devops right before merging! For example, touching envs/aws.py or or envs/common.py are dead giveaways that you'll need assistance when pushing your changes out. Things that require devops attention include, but are not limited to:
- configuration changes
- feature flags
- migrations
Ways to get devops attention, pick-and-choose whichever is most appropriate for your specific PR:
- [ ] Tag devops on your PR (`@edx/devops`), either for review (more specific to devops = better) or just a general FYI
- [ ] File a support ticket

FYI: Tag anyone who might be interested in this PR here.

## Post-review
- [ ] Squash commits

**Comment by andy-armstrong** [comment]
@robrap Here's my first attempted update to the APIs. I'm going to move on to update my other PR now, but I'd welcome your feedback here too (or you can wait to see the APIs in action).

**Comment by robrap** [comment]
@andy-armstrong Looks reasonable.  I need to get back to my docs.  The real question is how this play out in your other PRs.  Also, maybe @cahrens can help review the comments and make sure they are still correct?

File: src/js/utils/html-utils.js [hunk 1]
```diff
@@ -6,41 +6,75 @@
...
+         * @param {(string|HtmlSnippet)} html HTML provided either as
+         * a string or as an HtmlSnippet.
+         * @returns {HtmlSnippet} A safely escaped HTML snippet.
          */
-        escape = function(text) {
-            if (text.isHtmlSnippet) {
-                return text.toString();
+        ensureHtml = function(html) {
+            if (html instanceof HtmlSnippet) {
+                return html.toString();
```
Comment (cahrens) [inline src/js/utils/html-utils.js]: I don't understand the inconsistent return values of this method. In this case (html already an HtmlSnippet), this method returns a String. But in the case of a String being passed in, an HTMLSnippet is returned.

File: src/js/utils/html-utils.js [hunk 2]
```diff
@@ -109,15 +142,60 @@
...
-                    return rawTemplate(augmentedData);
+                    return HTML(rawTemplate(augmentedData));
                 };
             return template;
         };
 
+        /**
+         * A wrapper for $.html that safely escapes the provided HTML.
+         *
+         * If the HTML is provided as an HTML snippet then it is set directly
```
Comment (cahrens) [inline src/js/utils/html-utils.js]: I think you should include this explanation in the other two methods below as well.

File: src/js/utils/html-utils.js (hunk 1 above)
Comment (andy-armstrong) [inline src/js/utils/html-utils.js]: My bad. This should be returning an HtmlSnippet in both cases now.

Comment (andy-armstrong) [inline src/js/utils/html-utils.js]: I've fixed it and added a test.

File: src/js/utils/html-utils.js (hunk 2 above)
Comment (andy-armstrong) [inline src/js/utils/html-utils.js]: Fixed.

**Comment by andy-armstrong** [comment]
@robrap @cahrens This is ready for re-review. 

FYI @dan-f @AlasdairSwan

File: src/js/utils/html-utils.js [hunk 3]
```diff
@@ -6,41 +6,78 @@
...
         HTML = function(htmlString) {
-            var htmlSnippet = new String(htmlString); // jshint ignore:line
-            htmlSnippet.isHtmlSnippet = true;
-            return htmlSnippet;
+            return new HtmlSnippet(htmlString);
         };
 
         /**
-         * Returns an HTML string that is appropriately escaped.
+         * Ensures that the specified HTML is encoded as an HtmlSnippet.
```
Comment (robrap) [inline src/js/utils/html-utils.js]: Ensures that the provided text is properly HTML-escaped.

File: src/js/utils/html-utils.js [hunk 4]
```diff
@@ -6,41 +6,78 @@
...
          *
-         * Note: if the string provided is flagged as being an HTML snippet
+         * If the HTML is provided as a string, then it is HTML-escaped and
```
Comment (robrap) [inline src/js/utils/html-utils.js]: If a plain text string is provided, then it will be HTML-escaped and...

File: src/js/utils/html-utils.js [hunk 5]
```diff
@@ -6,41 +6,78 @@
...
+         * returned as an HtmlSnippet. If the parameter is an HTML snippet
          * then it will be returned directly so as not to double escape it.
          *
-         * @param {string} text A string that is either plan text, or is
-         * flagged as being an HTML snippet.
-         * @returns {string} A safely HTML-escaped string.
+         * @param {(string|HtmlSnippet)} html HTML provided either as
```
Comment (robrap) [inline src/js/utils/html-utils.js]:
```
* @param {(string|HtmlSnippet)} html Either a plain text string or an HTML snippet.
```

File: src/js/utils/html-utils.js [hunk 6]
```diff
@@ -6,41 +6,78 @@
...
+         * as a single root element.
+         *
+         * @param {HtmlSnippet} html The desired HTML provided as an HtmlSnippet.
+         * @returns {element} A new JQuery element.
+         */
+        element = function(html) {
+            if (!(html instanceof HtmlSnippet)) {
+                throw 'HtmlUtils.element received HTML that was not an HtmlSnippet: ' + html;
+            }
+            return $(ensureHtml(html).toString());
```
Comment (robrap) [inline src/js/utils/html-utils.js]: You could reverse the order of these and have an if/else, where the if doesn't require the "not".  Additionally, you don't need ensureHtml() if you are already checking that this must be an HtmlSnippet.

File: src/js/utils/html-utils.js [hunk 7]
```diff
@@ -6,41 +6,78 @@
...
+        };
+
+        /**
          * Returns an HTML snippet by interpolating the provided parameters.
          *
          * The HTML text is provided as a tokenized format string where parameters
```
Comment (robrap) [inline src/js/utils/html-utils.js]: The text is provided...

It's probably plain text that is being provided, not HTML.

File: src/js/utils/html-utils.js [hunk 8]
```diff
@@ -76,19 +113,39 @@
          *
          * @param {string} formatString The string to be interpolated.
          * @param {Object} parameters An optional set of parameters to the template.
```
Comment (robrap) [inline src/js/utils/html-utils.js]: The word "template" seems odd here.  Instead, maybe "...parameters for interpolation."

File: src/js/utils/html-utils.js [hunk 9]
```diff
@@ -107,17 +164,71 @@
...
 
+        /**
+         * A wrapper for $.html that safely escapes the provided HTML.
+         *
+         * If the HTML is provided as an HTML snippet then it is used directly.
+         * If the value is a string then it is assumed to be unescaped and
+         * so it is first escaped before being used.
+         *
+         * @param {element} element The element or elements to be updated.
+         * @param {(string|HtmlSnippet)} html The desired HTML, either as a string
```
Comment (robrap) [inline src/js/utils/html-utils.js]: ... either as a plain text string or an HTML snippet.

File: src/js/utils/html-utils.js [hunk 10]
```diff
@@ -107,17 +164,71 @@
...
+
+        /**
+         * A wrapper for $.append that safely escapes the provided HTML.
+         *
+         * If the HTML is provided as an HTML snippet then it is used directly.
+         * If the value is a string then it is assumed to be unescaped and
+         * so it is first escaped before being used.
+         *
+         * @param {element} element The element or elements to be updated.
+         * @param {(string|HtmlSnippet)} html The desired HTML, either as a string
```
Comment (robrap) [inline src/js/utils/html-utils.js]: Same as above.

File: src/js/utils/html-utils.js [hunk 11]
```diff
@@ -6,41 +6,78 @@
...
+         * A wrapper for $ that returns a new JQuery element for the specified HTML.
+         *
+         * Note that the HTML must have a single root element which will be the
+         * one returned. An error will be thrown if the HTML cannot be interpreted
+         * as a single root element.
+         *
+         * @param {HtmlSnippet} html The desired HTML provided as an HtmlSnippet.
+         * @returns {element} A new JQuery element.
+         */
+        element = function(html) {
```
Comment (robrap) [inline src/js/utils/html-utils.js]: Where and how is this method used?

File: src/js/utils/specs/html-utils-spec.js [hunk 12]
```diff
@@ -91,20 +114,131 @@ define(
...
+                    'escapes characters provided as strings': [
+                        ['Rock & Roll'],
+                        'Rock &amp; Roll'
+                    ],
+                    'does not escape HTML snippets': [
+                        [HtmlUtils.HTML('<a href="world">world</a>')],
+                        '<a href="world">world</a>'
+                    ],
+                    'can join a mixture of strings and HTML snippets': [
+                        ['Hello, ', HtmlUtils.HTML('<a href="world">world</a>')],
```
Comment (robrap) [inline src/js/utils/specs/html-utils-spec.js]: "Hello" should be "Rock & Roll" if you really want this to test a mix.

**Comment by cahrens** [comment]
Sorry, I don't have time to review today due to meetings and the fact that I have to leave at 2:45.

**Comment by andy-armstrong** [comment]
@robrap I've addressed your feedback if you could do a quick re-review.

@AlasdairSwan would you be able to review this as @cahrens won't be able to today.

File: src/js/utils/html-utils.js (hunk 11 above)
Comment (andy-armstrong) [inline src/js/utils/html-utils.js]: I've decided to remove this method for now so we can get this PR merged.

I was thinking we'd need it for calls to ``appendTo`` and ``prependTo`` where the caller has to create a new element from some HTML before attaching it. Here's one example:

https://github.com/edx/edx-platform/blob/master/lms/templates/courseware/progress_graph.js#L9

I'll revisit it when I have some more concrete examples where the existing methods won't be enough.

File: src/js/utils/html-utils.js [hunk 13]
```diff
@@ -75,20 +95,40 @@
...
+         * so it is first escaped before being joined.
+         *
+         * @param {...(string|HtmlSnippet)} items The strings and/or HTML snippets
+         * to be joined together.
+         * @returns {HtmlSnippet} The resulting safely escaped HTML snippet.
+         */
+        joinHtml = function() {
+            var html = '',
+                i;
+            for (i = 0; i < arguments.length; i++) {
```
Comment (AlasdairSwan) [inline src/js/utils/html-utils.js]: Nitpick: the length should be assigned to a variable instead of calculating length on each iteration.

**Comment by AlasdairSwan** [comment]
Just one minor nitpick @andy-armstrong . :+1:

File: src/js/utils/html-utils.js [hunk 14]
```diff
@@ -6,44 +6,64 @@
...
-    define(['underscore', 'edx-ui-toolkit/js/utils/string-utils'], function(_, StringUtils) {
-        var escape, interpolateHtml, HTML, template;
+    define(['underscore', 'jquery', 'edx-ui-toolkit/js/utils/string-utils'], function(_, $, StringUtils) {
+        var ensureHtml, interpolateHtml, joinHtml, HTML, template, setHtml, append, prepend;
 
         /**
-         * Helper function to return a string flagged as an HTML snippet.
+         * Creates an HTML snippet.
+         *
+         * The intention of an HTML snippet is to capture a string that
```
Comment (robrap) [inline src/js/utils/html-utils.js]: The intention of an HTML snippet is to communicate that the string it represents contains HTML that has been safely escaped as necessary.

File: src/js/utils/html-utils.js [hunk 15]
```diff
@@ -107,17 +147,70 @@
...
+         */
+        setHtml = function(element, html) {
+            return $(element).html(ensureHtml(html).toString());
+        };
+
+        /**
+         * A wrapper for $.append that safely escapes the provided HTML.
+         *
+         * If the HTML is provided as an HTML snippet then it is used directly.
+         * If the value is a string then it is assumed to be unescaped and
```
Comment (robrap) [inline src/js/utils/html-utils.js]: Are there ever real reasons to append a plain text string that gets escaped, or is this just a safety feature?  Same question for `setHtml()` above.  If it is just a safety feature, we might want to adjust the text to make that clear.

**Comment by robrap** [comment]
@andy-armstrong Minor comment and minor question.  I'll help you get to thumb as soon as you are available.

File: src/js/utils/html-utils.js (hunk 15 above)
Comment (andy-armstrong) [inline src/js/utils/html-utils.js]: See my comment on the other PR: https://github.com/edx/edx-platform/pull/11856/files#r57047568

IMO it is useful for ``setHtml`` and arguable for ``append``. I'd like to keep it for consistency and to enforce safety as you point out.

File: src/js/utils/html-utils.js (hunk 14 above)
Comment (andy-armstrong) [inline src/js/utils/html-utils.js]: Fixed.

File: src/js/utils/html-utils.js (hunk 15 above)
Comment (robrap) [inline src/js/utils/html-utils.js]: @andy-armstrong I was confirming and thinking the docs might want to make that clear, but I think it is fine and we'll see what happens.  Thanks.

**Comment by robrap** [comment]
:+1:   Thanks @andy-armstrong

**Pull request merged**